File: dairyTrack-web/src/Modules/controllers/nutritionController.js
// nutritionController.js
import { API_URL4 } from "../../api/apiController.js";
import Swal from "sweetalert2";

export const addNutrition = async (nutritionData) => {
  const user = JSON.parse(localStorage.getItem("user") || "{}");
  const token = user.token || null;
  console.log("addNutrition - Token:", token, "Data:", nutritionData);
  if (!token) {
    Swal.fire({
      icon: "error",
      title: "Sesi Berakhir",
      text: "Token tidak ditemukan. Silakan login kembali.",
    });
    localStorage.removeItem("user");
    window.location.href = "/";
    return { success: false, message: "Token tidak ditemukan." };
  }
  try {
    const response = await fetch(`${API_URL4}/nutrition`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(nutritionData),
    });

    const data = await response.json();
    console.log("addNutrition - Response:", { status: response.status, data });
    if (response.ok) {
      return { success: true, nutrition: data.data || data.nutrition || {}, message: data.message || "Nutrisi berhasil ditambahkan." };
    } else {
      return { success: false, message: data.message || "Gagal menambahkan nutrisi." };
    }
  } catch (error) {
    console.error("addNutrition - Error:", error.message || error);
    return {
      success: false,
      message: "Terjadi kesalahan saat menambahkan nutrisi.",
    };
  }
};

export const getNutritionById = async (nutritionId) => {
  const user = JSON.parse(localStorage.getItem("user") || "{}");
  const token = user.token || null;
  console.log("getNutritionById - Token:", token, "ID:", nutritionId);
  if (!token) {
    Swal.fire({
      icon: "error",
      title: "Sesi Berakhir",
      text: "Token tidak ditemukan. Silakan login kembali.",
    });
    localStorage.removeItem("user");
    window.location.href = "/";
    return { success: false, message: "Token tidak ditemukan." };
  }
  try {
    const response = await fetch(`${API_URL4}/nutrition/${nutritionId}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
    });

    const data = await response.json();
    if (response.status === 200) {
      return { success: true, nutrition: data.data };
    } else {
      return { success: false, message: data.message || "Nutrisi tidak ditemukan." };
    }
  } catch (error) {
    console.error("Error fetching nutrition by ID:", error);
    return {
      success: false,
      message: "Terjadi kesalahan saat mengambil data nutrisi.",
    };
  }
};

export const listNutritions = async () => {
  const user = JSON.parse(localStorage.getItem("user") || "{}");
  const token = user.token || null;
  console.log("listNutritions - Token:", token);
  if (!token) {
    Swal.fire({
      icon: "error",
      title: "Sesi Berakhir",
      text: "Token tidak ditemukan. Silakan login kembali.",
    });
    localStorage.removeItem("user");
    window.location.href = "/";
    return { success: false, message: "Token tidak ditemukan." };
  }
  try {
    const response = await fetch(`${API_URL4}/nutrition`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
    });

    const data = await response.json();
    if (response.status === 200) {
      return { success: true, nutritions: data.data };
    } else {
      return {
        success: false,
        message: data.message || "Gagal mengambil daftar nutrisi.",
      };
    }
  } catch (error) {
    console.error("Error fetching nutritions:", error);
    return {
      success: false,
      message: "Terjadi kesalahan saat mengambil daftar nutrisi.",
    };
  }
};

export const updateNutrition = async (nutritionId, nutritionData) => {
  const user = JSON.parse(localStorage.getItem("user") || "{}");
  const token = user.token || null;
  console.log("updateNutrition - Token:", token, "ID:", nutritionId);
  if (!token) {
    Swal.fire({
      icon: "error",
      title: "Sesi Berakhir",
      text: "Token tidak ditemukan. Silakan login kembali.",
    });
    localStorage.removeItem("user");
    window.location.href = "/";
    return { success: false, message: "Token tidak ditemukan." };
  }
  try {
    const response = await fetch(`${API_URL4}/nutrition/${nutritionId}`, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(nutritionData),
    });

    const data = await response.json();
    if (response.status === 200) {
      Swal.fire({
        icon: "success",
        title: "Sukses",
        text: data.message || "Nutrisi berhasil diperbarui.",
      });
      return { success: true, nutrition: data.data };
    } else {
      Swal.fire({
        icon: "error",
        title: "Gagal",
        text: data.message || "Gagal memperbarui nutrisi.",
      });
      return { success: false, message: data.message };
    }
  } catch (error) {
    Swal.fire({
      icon: "error",
      title: "Gagal",
      text: "Terjadi kesalahan saat memperbarui nutrisi.",
    });
    console.error("Error updating nutrition:", error);
    return {
      success: false,
      message: "Terjadi kesalahan saat memperbarui nutrisi.",
    };
  }
};

export const deleteNutrition = async (nutritionId) => {
  const user = JSON.parse(localStorage.getItem("user") || "{}");
  const token = user.token || null;
  console.log("deleteNutrition - Token:", token, "ID:", nutritionId);
  if (!token) {
    Swal.fire({
      icon: "error",
      title: "Sesi Berakhir",
      text: "Token tidak ditemukan. Silakan login kembali.",
    });
    localStorage.removeItem("user");
    window.location.href = "/";
    return { success: false, message: "Token tidak ditemukan." };
  }
  try {
    const response = await fetch(`${API_URL4}/nutrition/${nutritionId}`, {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
    });

    const data = await response.json();
    if (response.status === 200) {
      Swal.fire({
        icon: "success",
        title: "Sukses",
        text: data.message || "Nutrisi berhasil dihapus.",
      });
      return { success: true };
    } else {
      Swal.fire({
        icon: "error",
        title: "Gagal",
        text: data.message || "Gagal menghapus nutrisi.",
      });
      return { success: false, message: data.message };
    }
  } catch (error) {
    Swal.fire({
      icon: "error",
      title: "Gagal",
      text: "Terjadi kesalahan saat menghapus nutrisi.",
    });
    console.error("Error deleting nutrition:", error);
    return {
      success: false,
      message: "Terjadi kesalahan saat menghapus nutrisi.",
    };
  }
};

export const exportNutritionsToPDF = async () => {
  const user = JSON.parse(localStorage.getItem("user") || "{}");
  const token = user.token || null;
  console.log("exportNutritionsToPDF - Token:", token);
  if (!token) {
    Swal.fire({
      icon: "error",
      title: "Sesi Berakhir",
      text: "Token tidak ditemukan. Silakan login kembali.",
    });
    localStorage.removeItem("user");
    window.location.href = "/";
    return;
  }
  try {
    const response = await fetch(`${API_URL4}/nutrition/export/pdf`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
    });

    if (response.status === 200) {
      const blob = await response.blob();
      const url = window.URL.createObjectURL(blob);
      const a = document.createElement("a");
      a.href = url;
      a.download = "nutritions.pdf";
      document.body.appendChild(a);
      a.click();
      a.remove();
      Swal.fire({
        icon: "success",
        title: "Sukses",
        text: "Nutrisi berhasil diekspor ke PDF.",
      });
    } else {
      const error = await response.json();
      Swal.fire({
        icon: "error",
        title: "Gagal",
        text: error.message || "Gagal mengekspor nutrisi ke PDF.",
      });
    }
  } catch (error) {
    Swal.fire({
      icon: "error",
      title: "Gagal",
      text: "Terjadi kesalahan saat mengekspor nutrisi ke PDF.",
    });
    console.error("Error exporting nutritions to PDF:", error);
  }
};

export const exportNutritionsToExcel = async () => {
  const user = JSON.parse(localStorage.getItem("user") || "{}");
  const token = user.token || null;
  console.log("exportNutritionsToExcel - Token:", token);
  if (!token) {
    Swal.fire({
      icon: "error",
      title: "Sesi Berakhir",
      text: "Token tidak ditemukan. Silakan login kembali.",
    });
    localStorage.removeItem("user");
    window.location.href = "/";
    return;
  }
  try {
    const response = await fetch(`${API_URL4}/nutrition/export/excel`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
    });

    if (response.status === 200) {
      const blob = await response.blob();
      const url = window.URL.createObjectURL(blob);
      const a = document.createElement("a");
      a.href = url;
      a.download = "nutritions.xlsx";
      document.body.appendChild(a);
      a.click();
      a.remove();
      Swal.fire({
        icon: "success",
        title: "Sukses",
        text: "Nutrisi berhasil diekspor ke Excel.",
      });
    } else {
      const error = await response.json();
      Swal.fire({
        icon: "error",
        title: "Gagal",
        text: error.message || "Gagal mengekspor nutrisi ke Excel.",
      });
    }
  } catch (error) {
    Swal.fire({
      icon: "error",
      title: "Gagal",
      text: "Terjadi kesalahan saat mengekspor nutrisi ke Excel.",
    });
    console.error("Error exporting nutritions to Excel:", error);
  }
};

File: dairytrack_mobile/lib/controller/APIURL4/nutritionController.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../../api/apiController.dart';

class NutrisiManagementController {
  final String baseUrl = '$API_URL4/nutrition';
  final Duration _timeoutDuration = Duration(seconds: 10);

  Future<Map<String, String>> _getHeaders() async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('authToken');

    if (token == null) {
      throw Exception('No auth token found. Please login first.');
    }

    return {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer $token',
    };
  }

  Future<Map<String, dynamic>> addNutrisi(String name, String unit, int userId) async {
    try {
      final url = Uri.parse('$baseUrl');
      final headers = await _getHeaders();
      print('Sending add request: name=$name, unit=$unit, userId=$userId');
      final response = await http.post(
        url,
        headers: headers,
        body: jsonEncode({
          'name': name,
          'unit': unit,
          'user_id': userId,
          'created_by': userId,
          'updated_by': userId,
        }),
      ).timeout(_timeoutDuration, onTimeout: () {
        throw Exception('Request timed out. Please check your connection.');
      });

      print('Add Nutrisi Response: ${response.statusCode} - ${response.body}');

      if (response.statusCode >= 200 && response.statusCode < 300) {
        final responseData = jsonDecode(response.body);
        return {
          'success': true,
          'message': responseData['message'] ?? 'Nutrisi berhasil ditambahkan',
          'data': responseData['data'],
        };
      } else {
        try {
          final errorData = jsonDecode(response.body);
          String message = errorData['message'] ?? 'Failed to add nutrisi';
          if (message.contains('already exists') || message.contains('duplicate')) {
            message = 'Nutrisi dengan nama "$name" sudah ada. Silakan gunakan nama yang berbeda.';
          } else if (message.contains('user')) {
            message = 'User dengan ID $userId tidak ditemukan';
          } else if (message.contains('Token')) {
            message = 'Sesi Anda telah berakhir. Silakan login kembali.';
            await _logoutOnInvalidToken();
          }
          return {'success': false, 'message': message};
        } catch (e) {
          return {'success': false, 'message': 'Invalid response format: ${response.body}'};
        }
      }
    } catch (e) {
      print('Error adding nutrisi: $e');
      return {'success': false, 'message': 'An error occurred: $e'};
    }
  }

  Future<Map<String, dynamic>> updateNutrisi(int id, String name, String unit, int userId) async {
    try {
      final url = Uri.parse('$baseUrl/$id');
      final headers = await _getHeaders();
      print('Sending update request: id=$id, name=$name, unit=$unit, userId=$userId');
      final response = await http.put(
        url,
        headers: headers,
        body: jsonEncode({
          'name': name,
          'unit': unit,
          'updated_by': userId,
        }),
      ).timeout(_timeoutDuration, onTimeout: () {
        throw Exception('Request timed out. Please check your connection.');
      });

      print('Update Nutrisi Response: ${response.statusCode} - ${response.body}');

      if (response.statusCode >= 200 && response.statusCode < 300) {
        final responseData = jsonDecode(response.body);
        return {
          'success': true,
          'message': responseData['message'] ?? 'Nutrisi berhasil diperbarui',
          'data': responseData['data'],
        };
      } else {
        try {
          final errorData = jsonDecode(response.body);
          String message = errorData['message'] ?? 'Failed to update nutrisi';
          if (message.contains('already exists') || message.contains('duplicate')) {
            message = 'Nutrisi dengan nama "$name" sudah ada. Silakan gunakan nama yang berbeda.';
          } else if (message.contains('user')) {
            message = 'User dengan ID $userId tidak ditemukan';
          } else if (message.contains('Token')) {
            message = 'Sesi Anda telah berakhir. Silakan login kembali.';
            await _logoutOnInvalidToken();
          }
          return {'success': false, 'message': message};
        } catch (e) {
          return {'success': false, 'message': 'Invalid response format: ${response.body}'};
        }
      }
    } catch (e) {
      print('Error updating nutrisi: $e');
      return {'success': false, 'message': 'An error occurred: $e'};
    }
  }

  Future<Map<String, dynamic>> deleteNutrisi(int id) async {
    try {
      final url = Uri.parse('$baseUrl/$id');
      final headers = await _getHeaders();
      print('Sending delete request: id=$id');
      final response = await http.delete(
        url,
        headers: headers,
      ).timeout(_timeoutDuration, onTimeout: () {
        throw Exception('Request timed out. Please check your connection.');
      });

      print('Delete Nutrisi Response: ${response.statusCode} - ${response.body}');

      if (response.statusCode >= 200 && response.statusCode < 300) {
        final responseData = jsonDecode(response.body);
        return {
          'success': true,
          'message': responseData['message'] ?? 'Nutrisi berhasil dihapus',
        };
      } else {
        try {
          final errorData = jsonDecode(response.body);
          String message = errorData['message'] ?? 'Nutrisi tidak ditemukan';
          if (message.contains('Token')) {
            message = 'Sesi Anda telah berakhir. Silakan login kembali.';
            await _logoutOnInvalidToken();
          }
          return {'success': false, 'message': message};
        } catch (e) {
          return {'success': false, 'message': 'Invalid response format: ${response.body}'};
        }
      }
    } catch (e) {
      print('Error deleting nutrisi: $e');
      return {'success': false, 'message': 'An error occurred: $e'};
    }
  }

  Future<Map<String, dynamic>> getNutrisiById(int id) async {
    try {
      final url = Uri.parse('$baseUrl/$id');
      final headers = await _getHeaders();
      print('Sending get by id request: id=$id');
      final response = await http.get(
        url,
        headers: headers,
      ).timeout(_timeoutDuration, onTimeout: () {
        throw Exception('Request timed out. Please check your connection.');
      });

      print('Get Nutrisi by ID Response: ${response.statusCode} - ${response.body}');

      if (response.statusCode == 200) {
        final responseData = jsonDecode(response.body);
        return {
          'success': true,
          'data': responseData['data'],
        };
      } else {
        try {
          final errorData = jsonDecode(response.body);
          String message = errorData['message'] ?? 'Nutrisi tidak ditemukan';
          if (message.contains('Token')) {
            message = 'Sesi Anda telah berakhir. Silakan login kembali.';
            await _logoutOnInvalidToken();
          }
          return {'success': false, 'message': message};
        } catch (e) {
          return {'success': false, 'message': 'Invalid response format: ${response.body}'};
        }
      }
    } catch (e) {
      print('Error getting nutrisi by ID: $e');
      return {'success': false, 'message': 'An error occurred: $e'};
    }
  }

  Future<Map<String, dynamic>> getAllNutrisi() async {
    try {
      final url = Uri.parse('$baseUrl');
      final headers = await _getHeaders();
      print('Sending get all nutrisi request');
      final response = await http.get(
        url,
        headers: headers,
      ).timeout(_timeoutDuration, onTimeout: () {
        throw Exception('Request timed out. Please check your connection.');
      });

      print('Get All Nutrisi Response: ${response.statusCode} - ${response.body}');

      if (response.statusCode == 200) {
        final responseData = jsonDecode(response.body);
        return {
          'success': true,
          'data': responseData['data'],
        };
      } else {
        try {
          final errorData = jsonDecode(response.body);
          String message = errorData['message'] ?? 'Failed to get nutrisi';
          if (message.contains('Token')) {
            message = 'Sesi Anda telah berakhir. Silakan login kembali.';
            await _logoutOnInvalidToken();
          }
          return {'success': false, 'message': message};
        } catch (e) {
          return {'success': false, 'message': 'Invalid response format: ${response.body}'};
        }
      }
    } catch (e) {
      print('Error getting all nutrisi: $e');
      return {'success': false, 'message': 'An error occurred: $e'};
    }
  }

  Future<void> _logoutOnInvalidToken() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('authToken');
  }
}